### PR TITLE
perf: Rework symbol dependency structures

### PR DESCRIPTION
**Pull request by slavek-kucera**

### CODE REVIEW

File: parser_library/src/context/ordinary_assembly/symbol_dependency_tables.cpp [hunk 1]
```diff
@@ -221,6 +221,13 @@ struct dependant_visitor
     std::variant<id_index, space_ptr> operator()(space_ptr p) const { return std::move(p); }
 };
 
+struct dependant_hash
+{
+    size_t operator()(const id_index& id) const { return std::hash<id_index>()(id); }
```
Comment (jirimosinger) [inline parser_library/src/context/ordinary_assembly/symbol_dependency_tables.cpp:226]: Missing `#include <functional>`

File: parser_library/src/context/ordinary_assembly/symbol_dependency_tables.cpp [hunk 2]
```diff
@@ -450,13 +607,42 @@ bool symbol_dependency_tables::add_dependency(dependant target,
...
+
+    m_dependencies_iterators.pop_back();
+    m_dependencies_has_t_attr.pop_back();
+    m_dependencies_space_ptr_type.pop_back();
+    m_dependencies_filters.pop_back();
+
+    return std::move(m_dependencies.extract(it).key());
 }
 
+
```
Comment (jirimosinger) [inline parser_library/src/context/ordinary_assembly/symbol_dependency_tables.cpp:645]:
```suggestion
```

File: parser_library/src/context/ordinary_assembly/ordinary_assembly_context.h [hunk 3]
```diff
@@ -24,7 +24,7 @@
 #include "diagnostic_consumer.h"
 #include "section.h"
 #include "symbol.h"
-#include "symbol_dependency_tables.h"
+#include "tagged_index.h"
```
[the remark this answers is not in the document]
Comment (slavek-kucera) [reply parser_library/src/context/ordinary_assembly/ordinary_assembly_context.h:27]: `import std;` cannot get here fast enough :)

File: utils/include/utils/filter_vector.h [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+    {
+        const auto [bucket, bit] = deconstruct_value(v);
+        const T keep_on_mask = ~(one << bit);
+        const T summary_test_bit = one << bucket;
+
+        for (auto it = filters[bucket].begin(); auto& sum : filters[0])
+        {
+            auto& b = *it++;
+            if (!(sum & summary_test_bit))
+                continue;
```
[the remark this answers is not in the document]
Comment (slavek-kucera) [reply utils/include/utils/filter_vector.h:126]: nope, we need to go through the whole thing.

File: utils/include/utils/filter_vector.h [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+            if (acc.values[bucket] == 0)
+                continue;
+            const T keep_on_mask = ~acc.values[bucket];
+            const T summary_test_bit = one << bucket;
+
+            for (auto it = filters[bucket].begin(); auto& sum : filters[0])
+            {
+                auto& b = *it++;
+                if (!(sum & summary_test_bit))
+                    continue;
```
[the remark this answers is not in the document]
Comment (slavek-kucera) [reply utils/include/utils/filter_vector.h:146]: nope, the same reason, the only difference is that we are clearing all bits in the slice at the same time.

**Pull request merged**